modules/flowable-cmmn-engine/src/main/java/org/flowable/cmmn/engine/impl/interceptor/CmmnCommandInvoker.java

```diff
@@ -101,10 +104,30 @@ protected void executeOperation(CommandContext commandContext, boolean isStoreCa
 
     protected void evaluateUntilStable(CommandContext commandContext) {
         Set<String> involvedCaseInstanceIds = CommandContextUtil.getInvolvedCaseInstanceIds(commandContext);
-        if (involvedCaseInstanceIds != null) {
+        if (involvedCaseInstanceIds != null && !involvedCaseInstanceIds.isEmpty()) {
+            
+            CmmnEngineAgenda agenda = CommandContextUtil.getAgenda(commandContext);
 
             for (String caseInstanceId : involvedCaseInstanceIds) {
-                CommandContextUtil.getAgenda(commandContext).planEvaluateCriteriaOperation(caseInstanceId, true);
+                VariableListenerSession variableListenerSession = commandContext.getSession(VariableListenerSession.class);
+                Map<String, List<VariableListenerSessionData>> variableSessionData = variableListenerSession.getVariableData();
+                
+                if (variableSessionData != null) {
+                    List<String> variableListenerCaseInstanceIds = new ArrayList<>();
+                    for (String variableName : variableSessionData.keySet()) {
+                        List<VariableListenerSessionData> variableListenerDataList = variableSessionData.get(variableName);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: Iterating over [map.keySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#keySet--) and calling [map.get(Key)](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#get-java.lang.Object-) to get the values within the loop is inefficient. 
**Fix**: Iterate over [map.entrySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#entrySet--) and access the values using [entry.getValue()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.Entry.html#getValue--) to avoid potential performance penalty due to lookup.